<a href="https://colab.research.google.com/github/olegpant/python/blob/main/%D0%9A%D0%BE%D0%BF%D0%B8%D1%8F_%D0%B1%D0%BB%D0%BE%D0%BA%D0%BD%D0%BE%D1%82%D0%B0_%22hw_10_3_%D0%A0%D0%BE%D0%B7%D1%88%D0%B8%D1%80%D0%B5%D0%BD%D1%96_%D0%BC%D0%B5%D1%82%D0%BE%D0%B4%D0%B8_%D0%BE%D0%B1%D1%80%D0%BE%D0%B1%D0%BA%D0%B8_%D0%B4%D0%B0%D0%BD%D0%B8%D1%85_%D1%83_Pandas_ipynb%22.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Домашнє завдання до теми "Розширені методи обробки даних у Pandas"

В цьому домашньому завданні продовжуємо працювати з набором даних `supermarket_sales.csv`.

0. Імпортуйте бібліотеку pandas та зчитайте дані у змінну `df` типу `pandas.DataFrame`.

In [43]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [44]:
import pandas as pd

In [45]:
df = pd.read_csv('/content/drive/MyDrive/DataAnalystEdc/Pandas/data/supermarket_sales.csv')
df.head()

,Invoice ID,Branch,City,Customer type,Gender,Product line,Unit price,Quantity,Tax 5%,Total,Date,Time,Payment,cogs,gross margin percentage,gross income,Rating
0,750-67-8428,A,Yangon,Member,Female,Health and beauty,74.69,7,26.1415,548.9715,1/5/2019,13:08,Ewallet,522.83,4.761905,26.1415,9.1
1,226-31-3081,C,Naypyitaw,Normal,Female,Electronic accessories,15.28,5,3.8200,80.2200,3/8/2019,10:29,Cash,76.40,4.761905,3.8200,9.6
2,631-41-3108,A,Yangon,Normal,Male,Home and lifestyle,46.33,7,16.2155,340.5255,3/3/2019,13:23,Credit card,324.31,4.761905,16.2155,7.4
3,123-19-1176,A,Yangon,Member,Male,Health and beauty,58.22,8,23.2880,489.0480,1/27/2019,20:33,Ewallet,465.76,4.761905,23.2880,8.4
4,373-73-7910,A,Yangon,Normal,Male,Sports and travel,86.31,7,30.2085,634.3785,2/8/2019,10:37,Ewallet,604.17,4.761905,30.2085,5.3


1. Відсортуйте дані за спаданням ціни товара ('Unit price') і виведіть 5 перших рядочків на екран.

In [46]:
df.sort_values(by = 'Unit price', ascending = False).head(5)

,Invoice ID,Branch,City,Customer type,Gender,Product line,Unit price,Quantity,Tax 5%,Total,Date,Time,Payment,cogs,gross margin percentage,gross income,Rating
983,148-41-7930,C,Naypyitaw,Normal,Male,Health and beauty,99.96,7,34.986,734.706,1/23/2019,10:33,Cash,699.72,4.761905,34.986,6.1
122,219-22-9386,B,Mandalay,Member,Male,Sports and travel,99.96,9,44.982,944.622,3/9/2019,17:26,Credit card,899.64,4.761905,44.982,4.2
930,641-62-7288,B,Mandalay,Normal,Male,Home and lifestyle,99.92,6,29.976,629.496,3/24/2019,13:33,Ewallet,599.52,4.761905,29.976,7.1
494,437-53-3084,B,Mandalay,Normal,Male,Fashion accessories,99.89,2,9.989,209.769,2/26/2019,11:48,Ewallet,199.78,4.761905,9.989,7.1
283,667-92-0055,A,Yangon,Member,Male,Health and beauty,99.83,6,29.949,628.929,3/4/2019,15:02,Ewallet,598.98,4.761905,29.949,8.5


2. Знайдіть інвойс (рядок в даних) в категорії товарів `Electronic accessories`, в якому був куплений товар з найбільшою кількістю і найбільшою ціною. Виведіть цей рядок даних на екран.

In [47]:
df[df['Product line'] == 'Electronic accessories'].sort_values(by=['Quantity', 'Unit price'], ascending=False).head(1)

,Invoice ID,Branch,City,Customer type,Gender,Product line,Unit price,Quantity,Tax 5%,Total,Date,Time,Payment,cogs,gross margin percentage,gross income,Rating
105,704-48-3927,A,Yangon,Member,Male,Electronic accessories,88.67,10,44.335,931.035,1/12/2019,14:50,Ewallet,886.7,4.761905,44.335,7.3


In [48]:
df[df['Product line'] == 'Electronic accessories'].sort_values(by=['Quantity', 'Unit price'], ascending=False).iloc[0]

,105
Invoice ID,704-48-3927
Branch,A
City,Yangon
Customer type,Member
Gender,Male
Product line,Electronic accessories
Unit price,88.67
Quantity,10
Tax 5%,44.335
Total,931.035


3.1. Створіть нову колонку в даних `Rating Category`, яка визначає категорію рейтингу (`Rating`): високий, середній чи низький за наступним правилом:
- якщо рейтинг вищий або рівний 8 - то він високий (`High`)
- якщо рейтинг від 6 (включно) до 8 (не включно) - то він середній (`Medium`)
- інакше - низький (`Low`).

In [49]:
def rating_category(rating):
  if rating >=8:
    return 'High'
  elif rating >= 6:
    return 'Medium'
  else:
    return 'Low'

df['Rating Category'] = df['Rating'].apply(rating_category)

3.2. Виведіть, який відсоток даних становить кожне зі значень категорій рейтингу.

In [50]:
categories = df['Rating Category'].value_counts(normalize=True) * 100
categories

,proportion
Rating Category,
Medium,35.6
High,32.9
Low,31.5


4.1. Запустіть код нижче аби створити змінну `time`. Ця змінна містить час в форматі "година:хвилина". Напишіть функцію `get_hour`, яка буде зі змінної формату як `time` дістати лише години і конвертувати їх у тип `int`.

Запустіть написану функцію на змінній `time` та перевірте правильність роботи функції.

In [51]:
time = '11:28'

In [52]:
def get_hour(time):
  time = time.split(':')
  return int(time[0])


print(get_hour(time))

type(get_hour(time))



11


int

4.2. З допомогою функції `get_hour` і метода `DataFrame.apply()` створіть нову колонку `Hour` в нашому наборі даних, яка буде містити години здійснення покупки, обчислені на основі часу здійснення покупки `Time`.

In [53]:
df['Hour'] = df['Time'].apply(get_hour)
df[['Time', 'Hour']].head(10)

,Time,Hour
0,13:08,13
1,10:29,10
2,13:23,13
3,20:33,20
4,10:37,10
5,18:30,18
6,14:36,14
7,11:38,11
8,17:15,17
9,13:27,13


4.3. Виведіть кількість значень за кожною годиною покупки і дайте відповідь "О яких годинах найбільше здійснюють покупок покупці і чому на Ваш погляд саме о цих годинах? Чи це слогічно, що найбільше покупок саме в ці години?".

In [54]:
df['Hour'].value_counts().sort_index()

,count
Hour,
10,101
11,90
12,89
13,103
14,83
15,102
16,77
17,74
18,93


Найбільша кількість покупок спостерігається о 10 - 101 покупка, 13 - 103, 15:— 102 та 19 — 113.
Можливе пояснення таке: о 10-й годині покупки здійснюють люди, яким зручно відвідувати магазин у першій половині дня. Пік о 13-й годині може бути пов’язаний з обідньою перервою працівників, особливо якщо магазин розташований поруч з офісами або робочими районами.
О 15-й годині частина покупців може заходити до магазину після роботи або по дорозі додому. Найвищий показник о 19-й годині, ймовірно, пов’язаний з вечірнім часом, коли більшість людей уже завершила робочий день і має можливість зробити покупки.
Водночас варто враховувати масштаб різниці між годинами. Максимальний показник становить 113 покупок, а мінімальний — близько 74, тобто різниця приблизно у 1,5 раза. Це означає, що покупки протягом дня розподілені досить рівномірно, без різких сплесків або «мертвих» годин. Тому правильніше говорити не про один чіткий час пік, а про кілька помірних піків протягом дня.
Але щоб підтвердити причини таких піків, потрібно додатково врахувати розташування магазину, день тижня та тип покупців.


5. Виведіть середнє значення чеку (`Total`) для покупок в містах, наявних в даних, за допомогою функції `groupby`.

In [55]:
df.groupby('City')['Total'].mean()

,Total
City,
Mandalay,319.872506
Naypyitaw,337.099715
Yangon,312.354031


6. Виведіть кількість покупок, середнє, мінімальне та максимальне значення для загальної суми чеку (Total) в містах, наявних в даних, за допомогою функції groupby.
Зробіть висновки з цих показників та дайте рекомендації для мережі супермаркетів.

In [56]:
df.groupby('City')['Total'].agg(['count','mean', 'min', 'max'])


,count,mean,min,max
City,,,,
Mandalay,332,319.872506,18.6375,1022.49
Naypyitaw,328,337.099715,10.6785,1042.65
Yangon,340,312.354031,12.6945,1039.29


Найвищий середній чек має Naypyitaw — близько 337, тоді як у Mandalay він становить близько 320, а в Yangon — 312. Це показує, що для повного порівняння міст недостатньо аналізувати лише Yangon і Mandalay, оскільки Naypyitaw фактично є лідером за середнім чеком.
Кількість чеків у всіх трьох містах майже однакова, тому різниця в загальній виручці, скоріше за все, формується не через потік покупців, а через суму одного чека.
Для Yangon і Mandalay варто проаналізувати, які товари та категорії найчастіше купують у Naypyitaw, а також перевірити можливість збільшення середнього чека за допомогою крос-продажів, комплектів товарів, акцій та рекомендацій супутніх товарів.

7. Ми внесли деякі зміни в наш набір даних, тож, давайте запишемо результат у файл. Запишіть оновлену версію даних у файл `supermarket_sales_enriched.csv`. Зверніть увагу, аби не записати нічого зайвого, наприклад, індекси рядків нам в файлі не потрібні, оскільки вони стандартні.

In [57]:
df.to_csv('supermarket_sales_enriched.csv', index=False)